### PR TITLE
PERL-807: Implement Initial DNS Seedlist discovery spec

### PR DESCRIPTION
**Pull request by TBSliver**

### CODE REVIEW

File: lib/MongoDB/_URI.pm [hunk 1]
```diff
@@ -141,12 +156,166 @@ sub _parse_doc {
...
+    my %parsed;
+    for my $opt ( split '&', $result->{options} ) {
+        my @kv = split '=', $opt, -1;
+        MongoDB::UsageError->throw("expected key value pair") unless @kv == 2;
+        my ( $k, $v ) = map { _unescape_all($_) } @kv;
+        # connection string spec calls for case normalization
+        ( my $lc_k = $k ) =~ tr[A-Z][a-z];
+        if ( !$valid->{$lc_k} ) {
+            if ( $err_unsupported ) {
+                MongoDB::Error->throw("Unsupported option '$k' in URI $self\n");
```
Comment (xdg) [inline lib/MongoDB/_URI.pm]: Here, I think we'll be less confusing if we reported the TXT records.  "... in URI $self with TXT record $txt_record ..."

File: lib/MongoDB/_URI.pm [hunk 2]
```diff
@@ -141,12 +156,166 @@ sub _parse_doc {
...
+            } else {
+                warn "Unsupported option '$k' in URI $self\n";
+            }
+            next;
+        }
+        if ( exists $parsed{$lc_k} && !exists $options_with_list_type{$lc_k} ) {
+            warn "Multiple options were found for the same value '$lc_k'\n";
```
Comment (xdg) [inline lib/MongoDB/_URI.pm]: As long as I'm looking at this, we should have the warning message clarify that only the first one is used.

File: lib/MongoDB/_URI.pm [hunk 3]
```diff
@@ -141,12 +156,166 @@ sub _parse_doc {
     return $set;
 }
 
+use Devel::Dwarn;
```
[the remark this answers is not in the document]
Comment (TBSliver) [reply lib/MongoDB/_URI.pm]: ...oops

File: lib/MongoDB/_URI.pm [hunk 4]
```diff
@@ -141,12 +156,166 @@ sub _parse_doc {
...
+    my @hosts;
+    my $options = {};
+    my @split_name = split( '\.', $result{hostids} );
+    my $domain_name = join( '.', @split_name[1..$#split_name] );
+    if ( $srv_data ) {
+        foreach my $rr ( $srv_data->answer ) {
+            next unless $rr->type eq 'SRV';
+            my $target = $rr->target;
+            # search for dot before domain name for a valid hostname - can have sub-subdomain
+            unless ( $target =~ /\.${\$domain_name}$/ ) {
```
[the remark this answers is not in the document]
Comment (TBSliver) [reply lib/MongoDB/_URI.pm]: Ah this was just cos I was thinking in string interpolation mode and it worked. Will swap it over as its not a documented way of doing it as far as i can see in perlre

**Comment by TBSliver** [comment]
Changes pushed. Have left the checks for SRV and TXT records in as I am unsure without diving through most of the DNS specs to figure out if it will actually return anything BUT those in a request - however Net::DNS::RR does warn against not checking first: https://metacpan.org/pod/Net::DNS::RR#METHODS so am erring on the side of caution!

**Comment by xdg** [comment]
Squashed and committed as 32e90bc6a2b1660c0e3a610a8c175b55e60d81c0

**Pull request closed without merging**